### PR TITLE
[BugFix] Fix using wrong parquet column chunk index in dict filter

### PR DESCRIPTION
**Pull request by Smith-Cruise**
## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

We should not use `col_idx_in_parquet` to get ColumnChunkMetaData in `GroupReader::_process_columns_and_conjunct_ctxs()`, otherwise dictionary filter will use the wrong ColumnChunk metadata when the parquet schema contains complex types.

In this pr, I do the following two things:
* Using the correct index to get ColumnChunk metadata.
* Add some defense code in the parquet decoder.
* Refactor some function names.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/formats/parquet/group_reader.h [hunk 1]
```diff
@@ -35,10 +35,12 @@ namespace starrocks::parquet {
 struct GroupReaderParam {
     struct Column {
         // column index in parquet file
-        int col_idx_in_parquet;
+        int32_t col_idx_in_parquet;
+
+        int32_t column_chunk_idx_in_parquet;
```
Comment (dirtysalt) [inline be/src/formats/parquet/group_reader.h:40]: what's the difference with `col_idx_in_parquet` this field?

File: be/src/formats/parquet/group_reader.cpp [hunk 2]
```diff
@@ -215,7 +215,7 @@ void GroupReader::_process_columns_and_conjunct_ctxs() {
         int chunk_index = column.col_idx_in_chunk;
         SlotId slot_id = column.slot_id;
         const tparquet::ColumnMetaData& column_metadata =
-                _row_group_metadata->columns[column.col_idx_in_parquet].meta_data;
+                _row_group_metadata->columns[column.column_chunk_idx_in_parquet].meta_data;
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.cpp:218]:
```
const auto* schema_node = _param.file_metadata->schema().get_stored_column_by_field_idx(column.col_idx_in_parquet);
const auto column_chunk_idx_in_parquet = schema_node->physical_column_index;
```

**Review by DorianZheng — CHANGES_REQUESTED** [review]

File: be/src/formats/parquet/group_reader.h (hunk 1 above)
Comment (Smith-Cruise) [reply be/src/formats/parquet/group_reader.h:40]: for example parquet meta:
col a,
col b: {
   col: b1,
   col: b2
},
col c

then col c's col_idx_in_parquet = 2，column_chunk_idx_in_parquet = 3

File: be/src/formats/parquet/group_reader.h [hunk 3]
```diff
@@ -35,10 +35,12 @@ namespace starrocks::parquet {
 struct GroupReaderParam {
     struct Column {
         // column index in parquet file
-        int col_idx_in_parquet;
+        int32_t col_idx_in_parquet;
```
Comment (letian-jiang) [inline be/src/formats/parquet/group_reader.h:38]: suggest to use `field_idx_in_parquet`

Comment (Smith-Cruise) [reply be/src/formats/parquet/group_reader.h:38]: done

File: be/src/formats/parquet/group_reader.cpp (hunk 2 above)
Comment (Smith-Cruise) [reply be/src/formats/parquet/group_reader.cpp]: fixed

**Review by letian-jiang — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Comment by Smith-Cruise** [comment]
@Mergify branch-3.0

**Comment by Smith-Cruise** [comment]
@mergify backport branch-3.0

**Pull request merged**